### DISCUSSION TITLE
Chaining eager loading for custom relation.

### DISCUSSION THREAD

**Discussion by valikos** [jeremyevans/sequel#2035] [General]
I have some problem with chains of eager loading for custom associations.

One of my general association works well with chains of eager loading.

```
many_to_many :students, join_table: :lesson_students
``` 
So I can do `Lesson.eager(students: :profile).all` it will make 3 SQL requests. Everything in this case works as expected.
Or I can make something like
```
many_to_many :students, join_table: :lesson_students, eager: [:profile]
``` 
And this will do the same.

But with my custom association it doesn't want chain eager loading:

```
one_to_many :not_recorded_students,
            read_only: true,
            dataset_method: :not_recorded_students_dataset,
            dataset: (proc do
              Student.with(:not_recorded_students, Student.dataset.join(...).where(stud_id: id))
                           .from(:not_recorded_students)
            end),
            eager_loader: (lambda do |eo_opts|
              id_map = eo_opts[:id_map]
              eo_opts[:rows].each do |lesson|
                lesson.associations[:not_recorded_students] = []
              end
              students_dataset = Student.dataset.join(...).where(stud_id: id_map.keys)
              students_dataset = students_dataset.select_append { [Sequel[:lessons][:id].as(:planned_lesson_id)] }
              ds = Student.with(:not_recorded_students, students_dataset)
                                    .from(:not_recorded_students)
              ds.all do |student|
                recs = id_map[student.values.delete(:planned_lesson_id)]
                recs.each { |r| r.associations[:not_recorded_students] << student } if recs
              end
            end)
```
`Lesson.eager(not_recorded_students: :profile).all` will ignore eager loading for profiles.

I hope there is something that I am missing.

**Comment by valikos** [comment]
Looks like I found a way how to partly fix my problem.
```
            eager_loader: (lambda do |eo_opts|
              id_map = eo_opts[:id_map]
              eo_opts[:rows].each do |lesson|
                lesson.associations[:not_recorded_students] = []
              end
              students_dataset = Student.dataset.join(...).where(stud_id: id_map.keys).**eager(:profile)**
              students_dataset = students_dataset.select_append { [Sequel[:lessons][:id].as(:planned_lesson_id)] }
              ds.all do |student|
                recs = id_map[student.values.delete(:planned_lesson_id)]
                recs.each { |r| r.associations[:not_recorded_students] << student } if recs
              end
            end)
```
I added `eager(:profile)` to my custom dataset in eager_loader block. 
Chaining eager loading still is not working. Right now I reduced N+1 problem.

**Reply by jeremyevans** [reply]
As described in https://sequel.jeremyevans.net/rdoc/files/doc/advanced_associations_rdoc.html#label-Custom+Eager+Loaders, you probably want to use `eager(eo_opts[:associations])` to handle the recursive eager loading part.  If that isn't enough for want you want, can you put together a minimal self-contained example showing the problem?

**Reply by valikos** [reply]
I totally missed meaning of `eo_opts[:associations]`. 
This is was a missing puzzle in my case.
@jeremyevans thanks for advice.
